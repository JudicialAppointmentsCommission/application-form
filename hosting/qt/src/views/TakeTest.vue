<template>
  <main class="container">
    <h1>Take a qualifying test</h1>

    <div ref="loadingMessage" v-if="loaded === false && loadFailed === false">
      <span class="spinner-border spinner-border-sm text-secondary" aria-hidden="true"></span>
      Loading...
    </div>

    <div ref="errorMessage" v-if="loaded === false && loadFailed === true">
      <p>Could not load data. Please reload the page and try again.</p>
    </div>

    <div ref="qtView" v-if="loaded === true">
      <h2>{{qt.title}}</h2>
      <div v-if="qtHasOpened === false">
        <p>
          This qualifying test is currently closed.
          It'll open at <strong>{{qt.opening_time.toLocaleTimeString()}}</strong>
          on <strong>{{qt.opening_time.toLocaleDateString()}}</strong>.
        </p>
      </div>

      <div v-if="qtHasClosed === true">
        <p>This qualifying test has now closed.</p>
      </div>

      <div v-if="qtHasOpened && !qtHasClosed">
        <div v-if="allQtPhasesFinished">
          <p>You've completed all phases of this qualifying test.</p>
        </div>
        <div v-else>
          <p>You are taking the <strong>{{qt.title}}</strong> qualifying test.</p>
          <p>It consists of {{qt.phases.length}} phases.</p>
        </div>
        <TestPhase v-for="(phase, index) in qt.phases" :key="phase.title" :title="phase.title" :number="index+1" />
      </div>
    </div>
  </main>
</template>

<script>
  import { mapGetters } from 'vuex';
  import TestPhase from '@/components/TestPhase';

  export default {
    components: {
      TestPhase,
    },
    data() {
      return {
        qtId: 'ANEMLDHK21g6HtnXQBiC', // hardcoded for now
        loaded: false,
        loadFailed: false,
        isStarting: false,
      };
    },
    methods: {
      loadQtData() {
        return Promise.all([
          this.$store.dispatch('loadQt'),
          this.$store.dispatch('loadQtSummary'),
        ])
          .then(() => {
            this.loaded = true;
            this.$store.dispatch('subscribeQtSummary');
          })
          .catch(() => {
            this.loadFailed = true;
          });
      },
    },
    computed: {
      ...mapGetters([
        'qt',
        'qtHasOpened',
        'qtHasClosed',
        'allQtPhasesFinished',
      ]),
    },
    mounted() {
      this.$store.commit('setQtId', this.qtId);
      this.loadQtData();
    },
    destroyed() {
      this.$store.dispatch('unsubscribeQtSummary');
    },
  }
</script>
