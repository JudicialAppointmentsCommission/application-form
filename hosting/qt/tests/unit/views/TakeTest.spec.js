import {shallowMount} from "@vue/test-utils";
import TakeTest from '@/views/TakeTest';

const mockQt = {
  title: 'Combined QT',
  opening_time: new Date('2019-04-24 06:00'),
  closing_time: new Date('2019-04-24 21:00'),
  phases: [
    {
      title: 'Situational Judgement',
      form_url: 'https://qt.judicialappointments.digital/test/situational-judgement',
    },
    {
      title: 'Critical Analysis',
      form_url: 'https://qt.judicialappointments.digital/test/critical-analysis',
    }
  ],
};

/**
 * Helper function to collapse string whitespace (spaces & newlines) into single spaces.
 * @param string
 * @returns string
 */
const collapseWhitespace = (string) => {
  return string.replace(/[\s\n]+/g, ' ');
};

describe('views/TakeTest', () => {
  let store, wrapper;

  beforeEach(() => {
    store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: {
        qt: mockQt,
        qtHasOpened: false,
        qtHasClosed: false,
        qtPhaseFormUrl: null,
        qtPhaseCanBeStarted: null,
        qtPhaseHasBeenStarted: null,
        qtPhaseHasBeenFinished: null,
        allQtPhasesFinished: null,
      },
    };

    wrapper = shallowMount(TakeTest, {
      mocks: {
        $store: store,
      }
    });
  });

  describe.skip('lifecycle hooks', () => {
    describe('#mounted', () => {
      let mockLoadQtData;
      describe('initial loading state', () => {
        beforeEach(() => {
          mockLoadQtData = jest.fn();
          wrapper = shallowMount(TakeTest, {
            mocks: {
              $store: store,
            },
            methods: {
              loadQtData: mockLoadQtData,
            },
          });
        });

        it('renders "Loading..." text', () => {
          expect(wrapper.text()).toContain('Loading...');
        });

        it('renders an animated spinner', () => {
          const spinner = wrapper.find('.spinner-border');
          expect(spinner.exists()).toBe(true);
          expect(spinner.isVisible()).toBe(true);
        });

        it('commits the hardcoded Qualifying Test ID to the store', () => {
          const expectQtId = 'ANEMLDHK21g6HtnXQBiC';
          expect(store.commit).toHaveBeenCalledWith('setQtId', expectQtId);
        });

        it('executes `this.loadQtData`', () => {
          expect(mockLoadQtData).toHaveBeenCalled();
        });
      });

      /*it.each(['loadQt', 'loadQtSummary'])('dispatches "%s" to the store', (action) => {
        expect(store.dispatch).toHaveBeenCalledWith(action);
      });*/

      /*describe('when both "loadQt" and "loadQtSummary" Promises resolve', () => {
        beforeEach(() => {
          store.dispatch.mockResolvedValue();
        });
      });*/
    });
  });

  describe.skip('methods', () => {
    describe('#loadQtData', () => {
      it.each(['loadQt', 'loadQtSummary'])('dispatches "%s" to the Vuex store', (action) => {
        expect(store.dispatch).toHaveBeenCalledWith(action);
      });
    });
  });

  describe('template', () => {
    let loadingMessage, errorMessage, qtView;
    const findElements = () => {
      loadingMessage = wrapper.find({ref: 'loadingMessage'});
      errorMessage = wrapper.find({ref: 'errorMessage'});
      qtView = wrapper.find({ref: 'qtView'});
    };

    describe("when the QT data hasn't loaded yet", () => {
      beforeEach(() => {
        wrapper.setData({
          loaded: false,
          loadFailed: false,
        });
        findElements();
      });

      it('shows a loading message', () => {
        expect(loadingMessage.exists()).toBe(true);
        expect(loadingMessage.isVisible()).toBe(true);
      });

      it('loading message matches snapshot', () => {
        expect(loadingMessage.element).toMatchSnapshot();
      });

      it("doesn't show an error message", () => {
        expect(errorMessage.exists()).toBe(false);
      });

      it("doesn't show details of the QT", () => {
        expect(qtView.exists()).toBe(false);
      });
    });

    describe('when the QT data failed to load', () => {
      beforeEach(() => {
        wrapper.setData({
          loaded: false,
          loadFailed: true,
        });
        findElements();
      });

      it('shows an error message', () => {
        expect(errorMessage.exists()).toBe(true);
        expect(errorMessage.isVisible()).toBe(true);
      });

      it('error message matches snapshot', () => {
        expect(errorMessage.element).toMatchSnapshot();
      });

      it('no longer shows the loading message', () => {
        expect(loadingMessage.exists()).toBe(false);
      });

      it("doesn't show details of the QT", () => {
        expect(qtView.exists()).toBe(false);
      });
    });

    describe('when the QT data has loaded', () => {
      beforeEach(() => {
        store.getters.qt = mockQt;
        wrapper.setData({
          loaded: true,
          loadFailed: false,
        });
        findElements();
      });

      it('shows details of the QT', () => {
        expect(qtView.exists()).toBe(true);
      });

      it('no longer shows the loading message', () => {
        expect(loadingMessage.exists()).toBe(false);
      });

      it("doesn't show an error message", () => {
        expect(errorMessage.exists()).toBe(false);
      });

      it('shows the title of the QT', () => {
        expect(wrapper.find('h2').text()).toBe('Combined QT');
      });

      describe("the QT hasn't opened yet", () => {
        beforeEach(() => {
          store.getters.qtHasOpened = false;
          store.getters.qtHasClosed = false;
        });

        it('says the QT is closed', () => {
          expect(wrapper.text()).toContain('This qualifying test is currently closed.');
        });

        it('says when it will open', () => {
          const date = mockQt.opening_time.toLocaleDateString();
          const time = mockQt.opening_time.toLocaleTimeString();
          const expected = `It'll open at ${time} on ${date}`;
          const wrapperText = collapseWhitespace(wrapper.text());
          expect(wrapperText).toContain(expected);
        });
      });

      describe('the QT has already closed', () => {
        beforeEach(() => {
          store.getters.qtHasOpened = true;
          store.getters.qtHasClosed = true;
        });

        it('says the QT has closed', () => {
          expect(wrapper.text()).toContain('This qualifying test has now closed.');
        });
      });

      describe('the QT is open', () => {
        beforeEach(() => {
          store.getters.qtHasOpened = true;
          store.getters.qtHasClosed = false;
        });

        it('lists each phase of the test', () => {

        });

        describe('each test phase', () => {
          it('shows the phase title', () => {

          });


        });
      });
    });
  });

  describe('script', () => {

  });
});
