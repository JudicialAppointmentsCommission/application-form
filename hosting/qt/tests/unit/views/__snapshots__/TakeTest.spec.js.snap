// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/TakeTest template when the QT data failed to load error message matches snapshot 1`] = `
<div>
  <p>
    Could not load data. Please reload the page and try again.
  </p>
</div>
`;

exports[`views/TakeTest template when the QT data hasn't loaded yet loading message matches snapshot 1`] = `
<div>
  <span
    aria-hidden="true"
    class="spinner-border spinner-border-sm text-secondary"
  />
  
    Loading...
  
</div>
`;
